File smoosher changes

akashdw opened the pull request:
This pr is to support `FileSmoosher` changes proposed in https://github.com/druid-io/druid/issues/3513 .

`FileSmoosher` detect when it already has an OutputStream open and redirect newly opened OutputStreams to new files on the file system. When any of the open OutputStream objects are closed, they will also check to see if any of the other files have been closed in the meantime. If there are OutputStreams that have been closed, they will be copied on to the main smoosh file and the extra underlying file will be cleaned up. This has the downside of introducing an extra copy to these files, but it given that the strategy will only be used when absolutely necessary, this shouldn’t result in any noticeable performance degradation during indexing.

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -131,6 +136,11 @@ public void add(String name, List<ByteBuffer> bufferToAdd) throws IOException
   {
+    if(writerCurrentlyInUse)
+    {
+      return delegateSmooshedWriter (name , size);
```
weird spacing

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -64,8 +66,11 @@
   private final List<File> outFiles = Lists.newArrayList();
   private final Map<String, Metadata> internalFiles = Maps.newTreeMap();
+  private List<File> files = Lists.newArrayList();
+  private List<File> createFiles = Lists.newArrayList();
```
Should this be past-tense?

cheddar replied:
Actually, how about a rename to `filesInProcess`?

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
           throw new IOException(
               String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+              );
```
This whitespace appears to be off...

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
+              );
+        }
+
+        if(!writerCurrentlyInUse)
```
Given that you just set `writerCurrentlyInUse` to `false` and that this object is not attempting to be thread-safe, I don't think this can ever be false.

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
+        return addToOffset(Ints.checkedCast(java.nio.file.Files.copy(in, tmpFile.toPath())));
+      }
+
+      public int addToOffset(int numBytesWritten)
```
This method has a really weird contract.  

1. Why is it `public`?
2. Why does it return its incoming argument without any adjustment?

The name of the method makes it seem like this would be returning the total offset instead of the size.  The methods in the non-delegate writer all do this call first with the name "verify" which makes it clear that the intent of the method is to do some state checks...

akashdw replied:
Agree, It shouldn't be. Made the code changes to make it more meaningful.

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -64,8 +66,11 @@
 
   private final List<File> outFiles = Lists.newArrayList();
   private final Map<String, Metadata> internalFiles = Maps.newTreeMap();
+  private List<File> files = Lists.newArrayList();
```
How about renaming this to `completedFiles`.

akashdw replied:
👍

cheddar commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
+      {
+        open = false;
+        out.close();
+        files.add(tmpFile);
```
Why wouldn't we remove the file from the `createFiles` or the `filesInProcess` here?

cheddar commented on src/test/java/com/metamx/common/io/smoosh/SmooshedFileMapperTest.java:
```diff
@@ -50,25 +52,68 @@ public void testSanity() throws Exception
-      }
-    }
+  @Test
+  public void testWhenFirstWriterClosedInTheLast() throws Exception
```
Does "InTheLast" mean "AtTheEnd" or something else?

akashdw replied:
done

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -131,6 +136,11 @@ public void add(String name, List<ByteBuffer> bufferToAdd) throws IOException
   {
+    if(writerCurrentlyInUse)
+    {
+      return delegateSmooshedWriter(name , size);
```
space before ,

akashdw replied:
done

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -131,6 +136,11 @@ public void add(String name, List<ByteBuffer> bufferToAdd) throws IOException
 
   public SmooshedWriter addWithSmooshedWriter(final String name, final long size) throws IOException
   {
+    if(writerCurrentlyInUse)
```
Logic should go after argument check (if block below).

akashdw replied:
👍

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
+    }
+  }
+
+  private SmooshedWriter delegateSmooshedWriter(final String name,final long size) throws IOException
```
space after ,

akashdw replied:
done

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
+        completedFiles.add(tmpFile);
+        filesInProcess.remove(tmpFile);
+
+        if(!writerCurrentlyInUse ) {
```
spacing

akashdw replied:
fixed.

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
+  {	
+    final File tmpFile = new File(baseDir, name);
+    filesInProcess.add(tmpFile);
+    return new SmooshedWriter () {
```
no space before (), { on the new line

akashdw replied:
done

akashdw commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
+    //get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(completedFiles);
+    completedFiles = Lists.newArrayList();
+    for(File file: fileToProcess)
```
fixed

akashdw commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
   public void close() throws IOException
   {
+    //book keeping checks on created file.
+    if(!completedFiles.isEmpty())
```
fixed.

akashdw commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -195,13 +206,95 @@ public void close() throws IOException
+    //book keeping checks on created file.
+    if(!completedFiles.isEmpty())
+    {
+      for(File file: completedFiles)
```
done.

akashdw commented:
@logarithm @leventov @cheddar Thanks for reviewing the PR. I addressed all the comments and updated the pr. My Eclipse auto-formatter was turned off which caused the formatting issues.

cheddar commented:
👍

leventov commented:
Please don't merge until the end of this week, I want to review this.

akashdw commented:
@leventov Thanks for looking into this PR, I'm using https://github.com/druid-io/druid/blob/master/eclipse_formatting.xml to format code. I'll update my local settings if that setting got updated.
I updated the code and addressed all the formatting comments, not sure why github is not showing the updated code in comment section, however files changes section is showing formatted code.

akashdw commented:
@leventov Updated the PR and added comments to explain code logic.

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -63,8 +72,13 @@
 
   private final List<File> outFiles = Lists.newArrayList();
   private final Map<String, Metadata> internalFiles = Maps.newTreeMap();
+  // list of files completed writing content using delegated smooshedWriter.
```
I think it is better to write such comments as Javadoc `/** ... */`, then on automatic refactorings and reformattings IDEs will move comments along with fields. Javadoc comments are allowed for private fields.

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -206,13 +231,132 @@ public void close() throws IOException
+  private void mergeWithSmoosher() throws IOException
+  {
+    // Get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(completedFiles);
```
Maybe it can be just iteration over completedFiles (currently lines 251-256), then `completedFiles.clear()`.

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -206,13 +231,132 @@ public void close() throws IOException
+    {
+      for (File file: completedFiles)
+      {
+        file.delete();
```
Maybe merge completedFiles here using `mergeWithSmoosher()`, rather than silently delete?

leventov commented on src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java:
```diff
@@ -206,13 +231,132 @@ public void close() throws IOException
+      {
+        file.delete();
+      }
+      for (File file: filesInProcess)
```
Maybe throw some exception here, if filesToProcess is not empty? Because presence of files in `filesInProcess` means somebody opened a smoosherWriter but didn't close  it before closing FileSmoosher. It seems like a bug.

cheddar replied:
There is an exception thrown just as soon as this for loop exits for exactly that reason. It's a bug if any files haven't yet been merged by the time the larger Smoosher gets closed (each of the individual Writers should have been closed and those should have caused everything to get merged).

cheddar commented:
@leventov the idea to do it with composition is definitely interesting.  It poses a few problems as well because the only thing that currently has access to the `SmooshedWriter` to know if it is in use or not is the `FileSmoosher`.  So, the wrapping class would essentially also have to wrap the `SmooshedWriter` that is returned.

Also, there is a bit of state that the `FileSmoosher` necessarily owns, namely the directory that files are being created in.  Switching to composition would force this state to somehow be shared with the wrapping class (or the wrapping class would have to have yet another directory available to it for the storing of files).

I think that this code belongs inside `FileSmoosher` and going through the steps to make it work with composition will be just a bunch of code baggage without much benefit.

Can you elaborate on why you are afraid to change `FileSmoosher`?

leventov commented:
@cheddar 

> It poses a few problems as well because the only thing that currently has access to the SmooshedWriter to know if it is in use or not is the FileSmoosher. So, the wrapping class would essentially also have to wrap the SmooshedWriter that is returned.

From the description of https://github.com/druid-io/druid/issues/3513 I understood that @akashdw is going to pass FileSmoosher somewhere while SmoosherWriter is open. AFAIU it's not recursion, i. e. FileSmoosher is passed to some different place. A wrapper class could essentially have `delegateSmooshedWriter()` method public (however I would name it differently, somehow involving adjective "delayed"), and call this method from that other place. Don't need to track the FileSmoosher itself. I written this already in other words:

> I think this wrapper class could have specialized methods for writing to the delegate fileSmoosher directly and to temp files, because you know what do you need in the call sites in Druid. So the tricky writerCurrentlyInUse logic is not needed, I believe.

---

> Also, there is a bit of state that the FileSmoosher necessarily owns, namely the directory that files are being created in. Switching to composition would force this state to somehow be shared with the wrapping class (or the wrapping class would have to have yet another directory available to it for the storing of files).

The only bit of state that new logic accesses is baseDir, to create temp files. However temp files don't need (and actually shouldn't) be created in the smoosher directory itself. They should better be created via Java standard, or Guava's utility methods for creating temp files, which end up placing it in java.io.temp dir or /tmp, which is likely to be mounted to tmpfs, while the smoosher directory is less likely to be there (?)

---

So, I don't see the new logic is tightly coupled with the FileSmoosher state. Also a couple of weeks ago you mentioned that it's better to add code to druid than java-util. This also the case, I suppose.

leventov commented:
@cheddar Also, I think it's quite likely that  @akashdw's proposed design on adding functionality to Druid which he wants to add will be changes sooner or later. Maybe even as soon as on the PR review stage. So that this change is no longer needed, for whatever reason. But we will need to carry it because it's a library. While logic from Druid core could be removed pretty freely, because it is not a library. Also don't forget about boilerplate of having separate PRs and release cycle of java-util and Druid...

cheddar commented:
Ok, we will just move this to be part of the larger PR in druid.

cheddar commented:
Btw,

> They should better be created via Java standard, or Guava's utility methods for creating temp files, which end up placing it in java.io.temp dir or /tmp, which is likely to be mounted to tmpfs, while the smoosher directory is less likely to be there (?)

We cannot and should not do that.  By writing into random temporary locations, you are placing extra data in arbitrary places that the users do not have strong control over.  The directories used inside of Druid for storing these things have a very specific lifecycle and get cleaned up according to that lifecycle.  Reusing those directories is important and using tmp locations is a non-starter (and often leads to orphaned files and users not understanding why things have filled  up).

leventov commented:
@cheddar thanks, this makes sense. So, smooshed file directories are temporal as a whole? And could be located in tmpfs?

So if "MultiSmoosher" need to access FileSmoosher's baseDir, it could be passed along with FileSmoosher in constructor, or `getBaseDir()` accessor could be added to FileSmoosher.

drcrallen commented on src/test/java/com/metamx/common/io/smoosh/SmooshedFileMapperTest.java:
```diff
@@ -50,26 +52,71 @@ public void testSanity() throws Exception
+        if (i==10)
+        {
+          writer.write(ByteBuffer.wrap(Ints.toByteArray(19)));
+          CloseQuietly.close(writer);
```
why quietly?

drcrallen commented on src/test/java/com/metamx/common/io/smoosh/SmooshedFileMapperTest.java:
```diff
@@ -50,26 +52,71 @@ public void testSanity() throws Exception
+        smoosher.add(String.format("%d", i), tmpFile);
+        tmpFile.delete();
+      }
+      CloseQuietly.close(writer);
```
why quietly?

drcrallen commented on src/test/java/com/metamx/common/io/smoosh/SmooshedFileMapperTest.java:
```diff
@@ -150,4 +197,26 @@ public void testDeterministicFileUnmapping() throws IOException
+      }
+    }
+  }
+}
```
Missing newline

drcrallen commented:
Thanks!
